### PR TITLE
[GRAPHAPI] Implement MIOPEN_BACKEND_EXECUTION_PLAN_DESCRIPTOR Serialization

### LINKED ISSUE #2843: [MHA] Implement MIOPEN_BACKEND_OPERATIONGRAPH_DESCRIPTOR in Graph API
The following descriptors are to be implemented.

MIOPEN_BACKEND_OPERATIONGRAPH_DESCRIPTOR
  - [x] C functions and tests
  - [x] #2914

This descriptor should query pattern matcher for available solvers/engines capable of doing the requested calculations. The usercode chooses a solver/engine from the list, a read only attribute of the descriptor, and passes its index as an attribute to the next MIOPEN_BACKEND_ENGINE_DESCRIPTOR descriptor.

### LINKED ISSUE #2846: [MHA] Implement MIOPEN_BACKEND_EXECUTION_PLAN_DESCRIPTOR in Graph API
Implement MIOPEN_BACKEND_EXECUTION_PLAN_DESCRIPTOR

- [x] #2922
- [x] #2925
- [x] MIOPEN_BACKEND_EXECUTION_PLAN_DESCRIPTOR Serialization

### PR DESCRIPTION
**Pull request by scerzh**
closes #2846

and depends on #3428 which had not been merged before this PR was created

The PR consists of two parts:

* Serialization of `GraphExecutorFind20`
* Serialization of `ConvBiasResAddActivForwardExecutor`

### CODE REVIEW

**Review by DrizztDoUrden — APPROVED** [review]

File: src/graphapi/engine.cpp [hunk 1]
```diff
@@ -36,6 +43,179 @@ GraphPatternExecutor::~GraphPatternExecutor() = default;
 
 size_t GraphExecutorFind20::getWorkspaceSize() const { return mSolution.GetWorkspaceSize(); }
 
+namespace {
+
+const std::string_view base64Chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+                                     "abcdefghijklmnopqrstuvwxyz"
+                                     "0123456789+/";
+
+std::string base64Encode(const char* data, size_t length)
```
Comment (Vsevolod1983) [inline src/graphapi/engine.cpp]: Should we move base64 encode/decode to utils of any kind? Looks like it might be commonly used routine...

**Review by Vsevolod1983 — APPROVED** [review]

File: src/graphapi/engine.cpp (hunk 1 above)
Comment (CAHEK7) [reply src/graphapi/engine.cpp]: Why are the pointer and the size unbundled here? Can we pass `vector<char>`
Especially if it's inside an anonymous namespace and for a single specific case in that PR.

But probably for a common case, like Vsevolod mentioned, it's worth to have those fields separated.

UPD: you just wanted more convenient `length` usage.

File: src/graphapi/engine.cpp [hunk 2]
```diff
@@ -36,6 +43,179 @@ GraphPatternExecutor::~GraphPatternExecutor() = default;
...
+        while(i++ < 3)
+        {
+            encodedString += '=';
+        }
+    }
+
+    return encodedString;
+}
+
+std::string base64Decode(const std::string& encodedString)
```
Comment (CAHEK7) [inline src/graphapi/engine.cpp]: Returning binary blob as std::string may have some problems on Windows, I guess @apwojcik has already fixed a few cases and replaced them with a std::vector<char>.

But probably it was related to a serialization of such strings via `ofstream`. Could you double check it?

Comment (scerzh) [reply src/graphapi/engine.cpp]: Agree. `std::vector` would be better here, regardless of whether `std::string` has issues with Windows. Even better `std::vector<uint8_t>`, though requires a `reinterpret_cast` for `miopenLoadSolution` and `miopenSaveSolution`

File: src/graphapi/engine.cpp (hunk 1 above)
Comment (scerzh) [reply src/graphapi/engine.cpp]: Among possible options:
```C++
std::string base64Encode(const uint8_t* begin, const uint8_t* end)
std::string base64Encode(const uint8_t* begin, size_t size)
std::string base64Encode(const std::vector<uint8_t>& data)
```
I chose that one, but probably better to provide all of them.

**Pull request merged**